File: app/src/main/java/com/something15525/trimetgo/trimet_go/data/model/Route.java
package com.something15525.trimetgo.trimet_go.data.model;

public class Route {

    private String desc;
    private int routeId;
    private boolean isBus = false;
    private boolean isMax = false;
    private boolean isStreetcar = false;
    private boolean isWes = false;

    public String getDesc() {
        return this.desc;
    }

    public boolean isBus() {
        return this.isBus;
    }

    public boolean isMax() {
        return this.isMax;
    }

    public boolean isStreetcar() {
        return this.isStreetcar;
    }

    public boolean isWes() {
        return this.isWes;
    }

    public int getRouteId() {
        return this.routeId;
    }

    public String getTypeLetter() {
        if (this.isWes) {
            return "W";
        }
        if (this.isMax) {
            return "M";
        }
        if (this.isBus) {
            return "B";
        }
        return this.isStreetcar ? "S" : "Z";
    }

    public void setDesc(String desc) {
        this.desc = desc;
    }

    public void setRouteId(int routeId) {
        this.routeId = routeId;
    }

    public void setType(String type, String desc) {
        if (type.equals("R") && desc.contains("WES")) {
            this.isWes = true;
            return;
        }
        if (type.equals("R") && desc.contains("MAX")) {
            this.isMax = true;
        } else if (type.equals("B")) {
            this.isBus = true;
        }
    }

    public void setStreetcarType(String desc, String type) {
        if (desc.contains("Portland Streetcar") && type.equals("R")) {
            this.isStreetcar = true;
        }
    }
}

File: common/model/src/main/java/com/trimettransit/tracker/model/Route.kt
package com.trimettransit.tracker.model

/**
 * The single source of truth for the one-letter transit-type badges rendered in
 * route lists, stop markers, and trip-leg pills. Precedence for a route with
 * multiple flags: WES, MAX, bus, streetcar (see [Route.typeLetter]).
 */
object TransitTypeLetter {
    const val BUS = "B"
    const val MAX = "M"
    const val STREETCAR = "S"
    const val WES = "W"
    const val NONE = "Z"
}

data class Route(
    val desc: String = "",
    val routeId: Int = 0,
    val isBus: Boolean = false,
    val isMax: Boolean = false,
    val isStreetcar: Boolean = false,
    val isWes: Boolean = false
) {
    val typeLetter: String
        get() = when {
            isWes -> TransitTypeLetter.WES
            isMax || desc.contains("Vintage Trolley") -> TransitTypeLetter.MAX
            isBus -> TransitTypeLetter.BUS
            isStreetcar -> TransitTypeLetter.STREETCAR
            else -> TransitTypeLetter.NONE
        }
}
